File: packages/core/src/sqlkind.ts
/** Keyword-and-CTE-aware SQL kind. Not a regex over the whole string. */

import type { SqlKind } from "./types.ts";

const MUTATING = new Set(["INSERT", "UPDATE", "DELETE", "MERGE"]);
const DDL = new Set(["TRUNCATE", "DROP", "ALTER", "CREATE", "GRANT", "REINDEX", "VACUUM"]);

export interface SqlClassification {
  kind: SqlKind;
  table?: string;
  multiStatement: boolean;
  modifyingCte: boolean;
  forUpdate: boolean;
  returning: boolean;
}

export function stripSqlComments(sql: string): string {
  let out = "";
  let i = 0;
  let inS = false;
  let inD = false;
  let inLine = false;
  let inBlock = false;
  while (i < sql.length) {
    const c = sql[i]!;
    const n = sql[i + 1];
    if (inLine) {
      if (c === "\n") {
        inLine = false;
        out += c;
      }
      i++;
      continue;
    }
    if (inBlock) {
      if (c === "*" && n === "/") {
        inBlock = false;
        i += 2;
        continue;
      }
      i++;
      continue;
    }
    if (!inS && !inD && c === "-" && n === "-") {
      inLine = true;
      i += 2;
      continue;
    }
    if (!inS && !inD && c === "/" && n === "*") {
      inBlock = true;
      i += 2;
      continue;
    }
    if (!inD && c === "'") {
      inS = !inS;
      out += c;
      i++;
      continue;
    }
    if (!inS && c === '"') {
      inD = !inD;
      out += c;
      i++;
      continue;
    }
    out += c;
    i++;
  }
  return out;
}

function splitStatements(sql: string): string[] {
  const parts: string[] = [];
  let cur = "";
  let inS = false;
  let inD = false;
  for (let i = 0; i < sql.length; i++) {
    const c = sql[i]!;
    if (!inD && c === "'") inS = !inS;
    else if (!inS && c === '"') inD = !inD;
    if (c === ";" && !inS && !inD) {
      if (cur.trim()) parts.push(cur.trim());
      cur = "";
      continue;
    }
    cur += c;
  }
  if (cur.trim()) parts.push(cur.trim());
  return parts;
}

function nextKeyword(tokens: string[], i: number): string | undefined {
  return tokens[i]?.toUpperCase();
}

function tokenize(sql: string): string[] {
  const tokens: string[] = [];
  const re = /[A-Za-z_][\w$]*|"(?:[^"]|"")+"|\d+|./g;
  let m: RegExpExecArray | null;
  const s = sql.replace(/\s+/g, " ").trim();
  while ((m = re.exec(s))) {
    const t = m[0]!;
    if (t.trim() === "") continue;
    tokens.push(t);
  }
  return tokens;
}

function unquote(ident: string): string {
  if (ident.startsWith('"') && ident.endsWith('"')) return ident.slice(1, -1).replace(/""/g, '"');
  if (ident.startsWith("`") && ident.endsWith("`")) return ident.slice(1, -1);
  return ident;
}

function tableAfter(tokens: string[], keywords: string[]): string | undefined {
  for (let i = 0; i < tokens.length - 1; i++) {
    if (keywords.includes(tokens[i]!.toUpperCase())) {
      let j = i + 1;
      if (tokens[j]?.toUpperCase() === "ONLY") j++;
      if (tokens[j]?.toUpperCase() === "IF") return undefined;
      const t = tokens[j];
      if (t && /[A-Za-z_"]/.test(t[0]!)) return unquote(t.replace(/\.$/, ""));
    }
  }
  return undefined;
}

function hasModifyingCte(tokens: string[]): boolean {
  for (let i = 0; i < tokens.length; i++) {
    if (tokens[i]!.toUpperCase() === "AS" && tokens[i + 1] === "(") {
      for (let j = i + 2; j < Math.min(tokens.length, i + 12); j++) {
        if (MUTATING.has(tokens[j]!.toUpperCase())) return true;
        if (tokens[j] === ")") break;
      }
    }
  }
  return false;
}

function innermostMutating(tokens: string[]): SqlKind | undefined {
  let last: SqlKind | undefined;
  for (const t of tokens) {
    const u = t.toUpperCase();
    if (MUTATING.has(u) && u !== "MERGE") last = u as SqlKind;
  }
  return last;
}

function primaryKind(tokens: string[]): SqlKind {
  let i = 0;
  if (nextKeyword(tokens, i) === "WITH") {
    while (i < tokens.length && !["SELECT", "INSERT", "UPDATE", "DELETE", "TRUNCATE", "CALL", "EXEC"].includes(tokens[i]!.toUpperCase())) {
      i++;
    }
  }
  const k = nextKeyword(tokens, i);
  if (!k) return "UNKNOWN";
  if (k === "SELECT") return "SELECT";
  if (k === "INSERT") return "INSERT";
  if (k === "UPDATE") return "UPDATE";
  if (k === "DELETE") return "DELETE";
  if (k === "TRUNCATE") return "TRUNCATE";
  if (k === "DROP") return "DROP";
  if (k === "ALTER") return "ALTER";
  if (k === "CREATE") return "CREATE";
  if (k === "GRANT") return "GRANT";
  if (k === "REINDEX") return "REINDEX";
  if (k === "VACUUM") return "VACUUM";
  if (k === "CALL") return "CALL";
  if (k === "EXEC" || k === "EXECUTE") return "EXEC";
  if (DDL.has(k)) return k as SqlKind;
  return "UNKNOWN";
}

export function classifySql(sql: string): SqlClassification {
  const stripped = stripSqlComments(sql);
  const statements = splitStatements(stripped);
  const multiStatement = statements.length > 1;
  const primary = statements[0] ?? "";
  const tokens = tokenize(primary);
  const modifyingCte = hasModifyingCte(tokens);
  let kind = primaryKind(tokens);
  if (modifyingCte) {
    kind = innermostMutating(tokens) ?? kind;
  }
  const upper = tokens.map((t) => t.toUpperCase());
  const forUpdate =
    kind === "SELECT" &&
    (upper.includes("UPDATE") || upper.includes("SHARE")) &&
    upper.some((t, i) => t === "FOR" && (upper[i + 1] === "UPDATE" || upper[i + 1] === "SHARE"));
  const returning = upper.includes("RETURNING");
  let table: string | undefined;
  if (kind === "INSERT") table = tableAfter(tokens, ["INTO"]);
  else if (kind === "UPDATE") table = tableAfter(tokens, ["UPDATE"]);
  else if (kind === "DELETE") table = tableAfter(tokens, ["FROM"]);
  else if (kind === "SELECT") table = tableAfter(tokens, ["FROM"]);
  else if (kind === "TRUNCATE") table = tableAfter(tokens, ["TRUNCATE", "TABLE"]);
  else table = tableAfter(tokens, ["TABLE", "INDEX", "ON"]);
  return { kind, table, multiStatement, modifyingCte, forUpdate, returning };
}

File: sdk-python/src/latticeag_vekrevert/core/sqlkind.py
"""Keyword-and-CTE-aware SQL kind. Port of packages/core/src/sqlkind.ts."""

from __future__ import annotations

import re

MUTATING = {"INSERT", "UPDATE", "DELETE", "MERGE"}
DDL = {"TRUNCATE", "DROP", "ALTER", "CREATE", "GRANT", "REINDEX", "VACUUM"}


def strip_sql_comments(sql: str) -> str:
    out: list[str] = []
    i = 0
    in_s = in_d = in_line = in_block = False
    while i < len(sql):
        c = sql[i]
        n = sql[i + 1] if i + 1 < len(sql) else ""
        if in_line:
            if c == "\n":
                in_line = False
                out.append(c)
            i += 1
            continue
        if in_block:
            if c == "*" and n == "/":
                in_block = False
                i += 2
                continue
            i += 1
            continue
        if not in_s and not in_d and c == "-" and n == "-":
            in_line = True
            i += 2
            continue
        if not in_s and not in_d and c == "/" and n == "*":
            in_block = True
            i += 2
            continue
        if not in_d and c == "'":
            in_s = not in_s
            out.append(c)
            i += 1
            continue
        if not in_s and c == '"':
            in_d = not in_d
            out.append(c)
            i += 1
            continue
        out.append(c)
        i += 1
    return "".join(out)


def _split_statements(sql: str) -> list[str]:
    parts: list[str] = []
    cur = ""
    in_s = in_d = False
    for c in sql:
        if not in_d and c == "'":
            in_s = not in_s
        elif not in_s and c == '"':
            in_d = not in_d
        if c == ";" and not in_s and not in_d:
            if cur.strip():
                parts.append(cur.strip())
            cur = ""
            continue
        cur += c
    if cur.strip():
        parts.append(cur.strip())
    return parts


def _tokenize(sql: str) -> list[str]:
    return [t for t in re.findall(r'[A-Za-z_][\w$]*|"(?:[^"]|"")+"|\d+|.', sql.replace("\n", " ")) if t.strip()]


def _unquote(ident: str) -> str:
    if ident.startswith('"') and ident.endswith('"'):
        return ident[1:-1].replace('""', '"')
    if ident.startswith("`") and ident.endswith("`"):
        return ident[1:-1]
    return ident


def _table_after(tokens: list[str], keywords: list[str]) -> str | None:
    for i, t in enumerate(tokens[:-1]):
        if t.upper() in keywords:
            j = i + 1
            if j < len(tokens) and tokens[j].upper() == "ONLY":
                j += 1
            if j < len(tokens) and tokens[j].upper() == "IF":
                return None
            if j < len(tokens) and re.match(r'[A-Za-z_"]', tokens[j][0]):
                return _unquote(tokens[j].rstrip("."))
    return None


def _primary_kind(tokens: list[str]) -> str:
    i = 0
    if tokens and tokens[0].upper() == "WITH":
        while i < len(tokens) and tokens[i].upper() not in {
            "SELECT",
            "INSERT",
            "UPDATE",
            "DELETE",
            "TRUNCATE",
            "CALL",
            "EXEC",
        }:
            i += 1
    if i >= len(tokens):
        return "UNKNOWN"
    k = tokens[i].upper()
    if k in {
        "SELECT",
        "INSERT",
        "UPDATE",
        "DELETE",
        "TRUNCATE",
        "DROP",
        "ALTER",
        "CREATE",
        "GRANT",
        "REINDEX",
        "VACUUM",
        "CALL",
    }:
        return k
    if k in ("EXEC", "EXECUTE"):
        return "EXEC"
    if k in DDL:
        return k
    return "UNKNOWN"


def classify_sql(sql: str) -> dict:
    stripped = strip_sql_comments(sql)
    statements = _split_statements(stripped)
    multi = len(statements) > 1
    primary = statements[0] if statements else ""
    tokens = _tokenize(primary)
    upper = [t.upper() for t in tokens]
    modifying_cte = False
    for i, t in enumerate(tokens):
        if t.upper() == "AS" and i + 1 < len(tokens) and tokens[i + 1] == "(":
            for j in range(i + 2, min(len(tokens), i + 12)):
                if tokens[j].upper() in MUTATING:
                    modifying_cte = True
                    break
                if tokens[j] == ")":
                    break
    kind = _primary_kind(tokens)
    if modifying_cte:
        for t in tokens:
            u = t.upper()
            if u in MUTATING and u != "MERGE":
                kind = u
    for_update = (
        kind == "SELECT"
        and ("UPDATE" in upper or "SHARE" in upper)
        and any(upper[i] == "FOR" and i + 1 < len(upper) and upper[i + 1] in ("UPDATE", "SHARE") for i in range(len(upper)))
    )
    returning = "RETURNING" in upper
    table = None
    if kind == "INSERT":
        table = _table_after(tokens, ["INTO"])
    elif kind == "UPDATE":
        table = _table_after(tokens, ["UPDATE"])
    elif kind == "DELETE":
        table = _table_after(tokens, ["FROM"])
    elif kind == "SELECT":
        table = _table_after(tokens, ["FROM"])
    elif kind == "TRUNCATE":
        table = _table_after(tokens, ["TRUNCATE", "TABLE"])
    else:
        table = _table_after(tokens, ["TABLE", "INDEX", "ON"])
    return {
        "kind": kind,
        "table": table,
        "multiStatement": multi,
        "modifyingCte": modifying_cte,
        "forUpdate": for_update,
        "returning": returning,
    }
